### ISSUE TITLE
Bug: generateText is ignoring dynamic parameters due to a hard-coded model class

### ISSUE THREAD

**Issue by FWangZil** [elizaOS/eliza#1439]
**Is your feature request (or bug) related to a problem? Please describe.**

In the `generateText` and `generateMessageResponse` function, the `model class` is currently hard-coded(gpt-4o), which prevents dynamically switching to different models based on incoming parameters. This reduces flexibility and can increase development complexity in scenarios where different models are needed.

When using a non-OpenAI AI provider, an error is reported because the gpt-4o model cannot be found.

**Describe the solution you'd like**

- Remove the hard-coded `model class` reference in the `generateText` and `generateMessageResponse` function and instead use a model name/type passed in as a function parameter or configuration.
- Ensure there is a sensible default value to maintain compatibility with existing functionality or cases where a specific model is not provided.

**Describe alternatives you've considered**

- Leaving it as is would require manual code changes to switch models in different scenarios, which is not ideal.
- Introducing a separate function to handle switching model classes could lead to redundant logic and increase maintenance overhead.

**Additional context**

- **Risks**: This change is minor, but it’s important to check if other parts of the code depend on the old hard-coded logic.
- **Testing**:
  - Test with multiple model types to confirm that the function correctly switches to the specified model.

**Comment by FWangZil** [comment]
I tried to directly modify the hard-coded gpt-4o to the incoming ModelClass, but encountered a new problem. The current TikTok Model only supports a limited number of models in a list. This modification alone will not solve the problem, and may even introduce new problems.

**Comment by 0xCardinalError** [comment]
I had the same reverse problem actually.

![image](https://github.com/user-attachments/assets/95834a3e-7a10-4ae9-8965-45b6d8b26517)

was using anthropic model with generateObject and then it always failed. Worked only when I switched to using the hardcoded functions.  So as this is trimTokens that is problematic, seems to me that we should need to have some custom solution for triming. As https://www.npmjs.com/package/js-tiktoken  is openAI only and it will fail on each try to trim with other models.

**Comment by tcm390** [comment]
> I tried to directly modify the hard-coded gpt-4o to the incoming ModelClass, but encountered a new problem. The current TikTok Model only supports a limited number of models in a list. This modification alone will not solve the problem, and may even introduce new problems.

Hello! Could you clarify the issue you encountered with the TikTok Model supporting only a limited list of models? Are you referring to the trimTokens function?

**Comment by prince981620** [comment]
similar case is with twitter agents, i am passing api key and model name in character.json but it always fall back to one api key provided in .env (gemini api key) 
resulting api key gets exhausted while running multiple characters.

**Comment by FWangZil** [comment]
> > I tried to directly modify the hard-coded gpt-4o to the incoming ModelClass, but encountered a new problem. The current TikTok Model only supports a limited number of models in a list. This modification alone will not solve the problem, and may even introduce new problems.
> 
> Hello! Could you clarify the issue you encountered with the TikTok Model supporting only a limited list of models? Are you referring to the trimTokens function?

Yes, what I meant to say is Tiktoken. I'm sorry I wrote TikTok by mistake at first. I'm referring to the parameter `model: TiktokenModel` of the trimTokens method.

**Comment by impozzible** [comment]
@FWangZil I encountered the same issue today also, trying to use anthropic with generateObject, do you have a temporary fix for this?

**Comment by FWangZil** [comment]
> @FWangZil I encountered the same issue today also, trying to use anthropic with generateObject, do you have a temporary fix for this?

Sorry, I haven't found a particularly good method yet, because models other than openai seem to be outside the allowed range of TiktokenModel and need to be modified or processed in other ways.

**Comment by tcm390** [comment]
> > @FWangZil I encountered the same issue today also, trying to use anthropic with generateObject, do you have a temporary fix for this?
> 
> Sorry, I haven't found a particularly good method yet, because models other than openai seem to be outside the allowed range of TiktokenModel and need to be modified or processed in other ways.


Apologies for the delayed response; I was busy with another task. I've created a PR: https://github.com/elizaOS/eliza/pull/1605/. Could you test it to see if it resolves your issue?

I used the AutoTokenizer from Hugging Face's transformers library to handle non-tiktoken models. You'll need to set the environment variable for your custom tokenizer. You can find the model you need here: https://huggingface.co/models?sort=trending

For example:

env
```
TOKENIZER_MODEL="Xenova/claude-tokenizer"  
TOKENIZER_TYPE="auto"
```

**Comment by tcm390** [comment]
@impozzible, please kindly try the PR and see if it fixes your issue.

**Comment by tcm390** [comment]
Implemented in https://github.com/elizaOS/eliza/pull/1605

**Comment by prince981620** [comment]
> @impozzible, please kindly try the PR and see if it fixes your issue.

What about twitter agent with similar problem ??

**Comment by tcm390** [comment]
> > @impozzible, please kindly try the PR and see if it fixes your issue.
> 
> What about twitter agent with similar problem ??

Apologies, could you please create a new issue with more details, including steps to reproduce the problem and any relevant information?

**Comment by prince981620** [comment]
> > > @impozzible, please kindly try the PR and see if it fixes your issue.
> > 
> > 
> > What about twitter agent with similar problem ??
> 
> Apologies, could you please create a new issue with more details, including steps to reproduce the problem and any relevant information?

@impozzible please have a look at :
https://github.com/elizaOS/eliza/issues/1811#issue-2768816177

**Comment by mcnojo** [comment]
> similar case is with twitter agents, i am passing api key and model name in character.json but it always fall back to one api key provided in .env (gemini api key) resulting api key gets exhausted while running multiple characters.

Were you able to solve this?  I have a similar issue wherein the api key used for GenerateText is the one found in .env, not in character.settings.secrets.  If I dont include an api key in my .env then I get:

![Image](https://github.com/user-attachments/assets/9c56f412-e71a-4815-bcbc-0ae527664cd0)

**Comment by mcnojo** [comment]
Solved my specific problem, though the issue remains.  At its core:  The embedding system for messages is using the .env OPENAI_API_KEY, while the twitter client is using the secret.  Ie, functions from elizaos/core are being called form the twitter client, but using env vars to do text embedding, not the same key from character.settings.secrets
